controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp

```diff
@@ -46,14 +46,142 @@ dynamic_channel_selection_task::dynamic_channel_selection_task(db &database_,
 void dynamic_channel_selection_task::work()
 {
     switch (m_fsm_state) {
-    case eState::INIT:
-    case eState::IDLE:
-    case eState::TRIGGER_SCAN:
-    case eState::WAIT_FOR_SCAN_TRIGGERED:
-    case eState::WAIT_FOR_RESULTS_READY:
-    case eState::WAIT_FOR_RESULTS_DUMP:
-    case eState::SCAN_DONE:
-    case eState::ABORT_SCAN:
+    case eState::INIT: {
+        database.assign_dynamic_channel_selection_task_id(m_radio_mac, id);
+        // allow first scan
+        m_next_scan_timestamp_interval = std::chrono::steady_clock::now();
+        FSM_MOVE_STATE(IDLE);
+        break;
+    }
+    case eState::IDLE: {
+        if (m_is_single_scan_pending) {
+            m_is_single_scan         = true;
+            m_is_single_scan_pending = false;
+            FSM_MOVE_STATE(TRIGGER_SCAN);
+        } else if (database.get_channel_scan_is_enabled(m_radio_mac)) {
+            auto now = std::chrono::steady_clock::now();
+
+            if (now > m_next_scan_timestamp_interval) {
+                m_is_single_scan = false;
+                FSM_MOVE_STATE(TRIGGER_SCAN);
+                m_last_scan_try_timestamp = now;
+                LOG(DEBUG) << "interval condition is met, trigger scan";
+            }
+        }
+        break;
+    }
+    case eState::TRIGGER_SCAN: {
+        LOG(TRACE) << "TRIGGER_SCAN, mac=" << m_radio_mac << "scan_type is "
+                   << ((m_is_single_scan) ? "single-scan" : "continuous-scan");
+
+        // When a scan is requested send the scan parameters Channel pool & Dwell time
+        // Before sending the request set the scan_in_progress flag to true
+        // So another scan would not trigger on the same radio simultaneously
+
+        database.set_channel_scan_in_progress(m_radio_mac, true, m_is_single_scan);
```

Comment:
Why do we store this in the database? Is it accessed from additional places?

Reply:
yes, we cannot allow user to trigger chan scan on the same radio in the middle of a already running scan

Reply:
check if can read task fsm state instead #665